File: common/src/test/java/com/taskadapter/connector/common/DefaultValueSetterTest.java
package com.taskadapter.connector.common;

import com.taskadapter.connector.FieldRow;
import com.taskadapter.connector.definition.TaskId;
import com.taskadapter.model.AllFields;
import com.taskadapter.model.CustomDate;
import com.taskadapter.model.CustomFloat;
import com.taskadapter.model.CustomSeqString;
import com.taskadapter.model.CustomString;
import com.taskadapter.model.DateTypeTag;
import com.taskadapter.model.Field;
import com.taskadapter.model.GTask;
import org.junit.Test;
import scala.collection.JavaConverters;
import scala.collection.mutable.Buffer;

import java.text.ParseException;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;

public class DefaultValueSetterTest {
    static List<FieldRow<?>> defaultRows = List.of(
            new FieldRow(Optional.of(AllFields.summary()), Optional.of(AllFields.summary()), ""),
            new FieldRow(Optional.of(AllFields.description()), Optional.of(AllFields.description()), "")
    );

    @Test
    public void taskIsDeepCloned() {
        var originalTask = new GTask()
                .setValue(AllFields.description(), "original description");
        var newTask = DefaultValueSetter.adapt(defaultRows, originalTask);
        originalTask.setValue(AllFields.description(), "new description");
        assertThat(newTask.getValue(AllFields.description())).isEqualTo("original description");
    }

    @Test
    public void defaultValueIsSetIfSourceFieldValueIsEmpty() {
        var rows = rows(
                new FieldRow<>(Optional.of(AllFields.summary()), Optional.of(AllFields.summary()), ""),
                new FieldRow<>(Optional.of(AllFields.description()), Optional.of(AllFields.description()), "default description")
        );
        var originalTask = new GTask()
                .setValue(AllFields.description(), "");

        var newTask = DefaultValueSetter.adapt(rows, originalTask);
        assertThat(newTask.getValue(AllFields.description())).isEqualTo("default description");
    }


    // TA supports one-sided mappings where a default value is provided to the target Field and there is
    // no corresponding Field in the source connector
    @Test
    public void defaultValueIsSetIfSourceFieldIsNotDefinedButDefaultValueExists() {
        var rows = rows(
                new FieldRow(Optional.empty(), Optional.of(AllFields.description()), "default description")
        );
        var originalTask = new GTask()
                .setValue(AllFields.description(), "");

        var newTask = DefaultValueSetter.adapt(rows, originalTask);
        assertThat(newTask.getValue(AllFields.description())).isEqualTo("default description");
    }

    @Test
    public void fieldIsSafelySkippedIfTargetFieldIsNotDdefined() {
        var rows = rows(
                new FieldRow(Optional.of(AllFields.description()), Optional.empty(), "default"));
        var newTask = DefaultValueSetter.adapt(rows,
                new GTask().setValue(AllFields.description(), ""));
        assertThat(newTask.getValue(AllFields.description())).isNull();
    }

    @Test
    public void existingValueIsPreservedWhenFieldHasIt() {
        var rows = rows(
                new FieldRow(Optional.of(AllFields.summary()), Optional.of(AllFields.summary()), ""),
                new FieldRow(Optional.of(AllFields.description()), Optional.of(AllFields.description()), "default description")
        );
        var originalTask = new GTask()
                .setValue(AllFields.description(), "something");
        var newTask = DefaultValueSetter.adapt(rows, originalTask);
        assertThat(newTask.getValue(AllFields.description())).isEqualTo("something");
    }

    // without this creating subtasks won't work, at least in JIRA
    @Test
    public void parentKeyIsPreserved() {
        var task = new GTask();
        var identity = new TaskId(1L, "parent1");
        task.setParentIdentity(identity);
        var newTask = DefaultValueSetter.adapt(defaultRows, task);
        assertThat(newTask.getParentIdentity()).isEqualTo(identity);
    }

    // regression test for https://bitbucket.org/taskadapter/taskadapter/issues/85/subtasks-are-not-saved
    @Test
    public void childrenArePreserved() {
        var parent = new GTask()
                .setId(1L);
        var sub = new GTask()
                .setId(100L);
        parent.addChildTask(sub);

        var adapted = DefaultValueSetter.adapt(defaultRows, parent);
        assertThat(adapted.getChildren()).hasSize(1);
    }

    @Test
    public void dateFieldTypeIsAdapted() {
        var rows = rows(
                new FieldRow(Optional.of(AllFields.dueDate()), Optional.of(AllFields.dueDate()), null)
        );
        var task = new GTask();
        var date = new Date();
        task.setValue(AllFields.dueDate(), date);
        var newTask = DefaultValueSetter.adapt(rows, task);
        assertThat(newTask.getValue(AllFields.dueDate())).isEqualTo(date);
    }

    @Test
    public void setsDefaultValuesWithProperTypes() {
        checkField(AllFields.assigneeLoginName(), "login", "login");
        checkField(AllFields.assigneeFullName(), "name", "name");
        checkField(AllFields.priority(), "1", 1);
        checkDate(AllFields.closedOn(), "2018 05 04");
        checkField(AllFields.components(), "c1 c2", scalaSeq("c1", "c2"));
        checkField(AllFields.components(), "c1", scalaSeq("c1"));
        checkDate(CustomDate.apply("date1"), "2018 05 04");
        checkField(CustomFloat.apply("float"), "1.2", 1.2f);
        checkField(CustomSeqString.apply("elements"), "a b", scalaSeq("a", "b"));
        checkField(CustomString.apply("custom1"), "text", "text");
        checkDate(AllFields.createdOn(), "2018 05 04");
        checkField(AllFields.description(), "text", "text");
        checkDate(AllFields.dueDate(), "2018 05 04");
        checkField(AllFields.dueDate(), "", null);
        checkField(AllFields.doneRatio(), "33", 33F);
        checkField(AllFields.estimatedTime(), "10.5", 10.5f);
//    checkField(SpentTime, "3.3", 3.3f);
        checkField(AllFields.id(), "5", 5L);
        checkField(AllFields.key(), "TEST-1", "TEST-1");
        checkDate(AllFields.startDate(), "2018 05 04");
        checkField(AllFields.summary(), "text", "text");
        checkField(AllFields.targetVersion(), "1.0", "1.0");
        checkField(AllFields.taskStatus(), "new", "new");
        checkField(AllFields.taskType(), "feature", "feature");
        checkField(AllFields.reporterLoginName(), "login", "login");
        checkField(AllFields.reporterFullName(), "name", "name");
        checkDate(AllFields.updatedOn(), "2018 05 04");
    }

    private Buffer<String> scalaSeq(String... values) {
        return JavaConverters.asScalaBuffer(List.of(values));
    }


    private static void checkDate(Field<?> field, String str) {
        try {
            checkField(field, str, DateTypeTag.DATE_PARSER().parse(str));
        } catch (ParseException e) {
            throw new RuntimeException(e);
        }
    }

    private static void checkField(Field<?> field, String defaultString, Object expectedValue) {
        var rows = rows(
                new FieldRow(Optional.of(field), Optional.of(field), defaultString));
        var newTask = DefaultValueSetter.adapt(rows, new GTask());
        assertThat(newTask.getValue(field)).isEqualTo(expectedValue);
    }


    private static List<FieldRow<?>> rows(FieldRow<?>... rows) {
        return Arrays.asList(rows.clone());
    }
}

File: common/src/test/scala/com/taskadapter/connector/common/DefaultValueSetterTest.scala
package com.taskadapter.connector.common

import java.util.Date

import com.taskadapter.connector.FieldRow
import com.taskadapter.connector.definition.TaskId
import com.taskadapter.model._
import org.fest.assertions.Assertions.assertThat
import org.junit.runner.RunWith
import org.scalatest.concurrent.ScalaFutures
import org.scalatest.junit.JUnitRunner
import org.scalatest.{FunSpec, Matchers}

@RunWith(classOf[JUnitRunner])
class DefaultValueSetterTest extends FunSpec with ScalaFutures with Matchers {
  val defaultRows = List(
    FieldRow(Summary, Summary, ""),
    FieldRow(Description, Description, "")
  )

  it("task is deep cloned") {
    val originalTask = new GTask
    originalTask.setValue(Description, "original description")

    val newTask = DefaultValueSetter.adapt(defaultRows, originalTask)
    originalTask.setValue(Description, "new description")
    assertThat(newTask.getValue(Description)).isEqualTo("original description")

  }

  it("default value is set if source field value is empty") {
    val rows = List(
      FieldRow(Summary, Summary, ""),
      FieldRow(Description, Description, "default description")
    )
    val originalTask = new GTask
    originalTask.setValue(Description, "")

    val newTask = DefaultValueSetter.adapt(rows, originalTask)
    newTask.getValue(Description) shouldBe "default description"
  }

  // TA supports one-sided mappings where a default value is provided to the target Field and there is
  // no corresponding Field in the source connector
  it("default value is set if source field is not defined but default value exists") {
    val rows = List(
      new FieldRow(None, Some(Description), "default description")
    )
    val originalTask = new GTask
    originalTask.setValue(Description, "")

    val newTask = DefaultValueSetter.adapt(rows, originalTask)
    newTask.getValue(Description) shouldBe "default description"
  }

  it("field is safely skipped if target field is not defined") {
    val rows = List(new FieldRow(Some(Description), None, "default"))
    val newTask = DefaultValueSetter.adapt(rows,
      new GTask().setValue(Description, ""))
    newTask.getValue(Description) shouldBe null
  }

  it("existing value is preserved when field has it") {
    val rows = List(
      FieldRow(Summary, Summary, ""),
      FieldRow(Description, Description, "default description")
    )
    val originalTask = new GTask
    originalTask.setValue(Description, "something")
    val newTask = DefaultValueSetter.adapt(rows, originalTask)
    newTask.getValue(Description) shouldBe "something"
  }

  // without this creating subtasks won't work, at least in JIRA
  it("parent key is preserved") {
    val task = new GTask
    val identity = TaskId(1, "parent1")
    task.setParentIdentity(identity)
    val newTask = DefaultValueSetter.adapt(defaultRows, task)
    newTask.getParentIdentity shouldBe identity
  }

  // regression test for https://bitbucket.org/taskadapter/taskadapter/issues/85/subtasks-are-not-saved
  it("children are preserved") {
    val parent = new GTask
    parent.setId(1l)
    val sub = new GTask
    sub.setId(100l)
    parent.addChildTask(sub)

    val adapted = DefaultValueSetter.adapt(defaultRows, parent)
    adapted.getChildren.size() shouldBe 1
  }

  it("Date field type is adapted") {
    val rows = List(
      FieldRow(DueDate, DueDate, null)
    )
    val task = new GTask
    val date = new Date
    task.setValue(DueDate, date)
    val newTask = DefaultValueSetter.adapt(rows, task)
    newTask.getValue(DueDate) shouldBe date
  }

  it("sets default values with proper types") {
    checkField(AssigneeLoginName, "login", "login")
    checkField(AssigneeFullName, "name", "name")
    checkField(Priority, "1", 1)
    checkDate(ClosedOn, "2018 05 04")
    checkField(Components, "c1 c2", Seq("c1", "c2"))
    checkField(Components, "c1", Seq("c1"))
    checkDate(CustomDate("date1"), "2018 05 04")
    checkField(CustomFloat("float"), "1.2", 1.2f)
    checkField(CustomSeqString("elements"), "a b", Seq("a", "b"))
    checkField(CustomString("custom1"), "text", "text")
    checkDate(CreatedOn, "2018 05 04")
    checkField(Description, "text", "text")
    checkDate(DueDate, "2018 05 04")
    checkField(DueDate, "", null)
    checkField(DoneRatio, "33", 33)
    checkField(EstimatedTime, "10.5", 10.5f)
    checkField(SpentTime, "3.3", 3.3f)
    checkField(Id, "5", 5)
    checkField(Key, "TEST-1", "TEST-1")
    checkDate(StartDate, "2018 05 04")
    checkField(Summary, "text", "text")
    checkField(TargetVersion, "1.0", "1.0")
    checkField(TaskStatus, "new", "new")
    checkField(TaskType, "feature", "feature")
    checkField(ReporterLoginName, "login", "login")
    checkField(ReporterFullName, "name", "name")
    checkDate(UpdatedOn, "2018 05 04")
  }

  private def checkDate(field: Field[_], str: String): Unit = {
    checkField(field, str, DateTypeTag.DATE_PARSER.parse(str))
  }

  private def checkField(field: Field[_], defaultString: String, expectedValue: Any): Unit = {
    val rows = Seq(FieldRow(field.asInstanceOf[Field[Any]], field.asInstanceOf[Field[Any]], defaultString))
    val newTask = DefaultValueSetter.adapt(rows, new GTask)
    newTask.getValue(field) shouldBe expectedValue
  }

}


